Gems/ROS2/Code/Source/Lidar/ROS2LidarSensorComponent.cpp

```diff
@@ -139,89 +140,70 @@ namespace ROS2
+
+        builder.AddField("x", sensor_msgs::msg::PointField::FLOAT32)
+            .AddField("y", sensor_msgs::msg::PointField::FLOAT32)
+            .AddField("z", sensor_msgs::msg::PointField::FLOAT32);
 
-        sensor_msgs::PointCloud2Modifier modifier(message);
         if (isIntensityEnabled)
         {
-            modifier.setPointCloud2Fields(
-                4,
-                "x",
-                1,
-                sensor_msgs::msg::PointField::FLOAT32,
-                "y",
-                1,
-                sensor_msgs::msg::PointField::FLOAT32,
-                "z",
-                1,
-                sensor_msgs::msg::PointField::FLOAT32,
-                "intensity",
-                1,
-                sensor_msgs::msg::PointField::FLOAT32);
+            builder.AddField("intensity", sensor_msgs::msg::PointField::FLOAT32);
         }
-        else
-        {
-            modifier.setPointCloud2Fields(
-                3,
-                "x",
-                1,
-                sensor_msgs::msg::PointField::FLOAT32,
-                "y",
-                1,
-                sensor_msgs::msg::PointField::FLOAT32,
-                "z",
-                1,
-                sensor_msgs::msg::PointField::FLOAT32);
-        }
-        modifier.resize(results.m_points.size());
+        sensor_msgs::msg::PointCloud2 message = builder.Build();
+
+
+        sensor_msgs::PointCloud2Iterator<float> messageXIt(message, "x");
+        sensor_msgs::PointCloud2Iterator<float> messageYIt(message, "y");
+        sensor_msgs::PointCloud2Iterator<float> messageZIt(message, "z");
 
-        sensor_msgs::PointCloud2Iterator<float> xIt(message, "x");
-        sensor_msgs::PointCloud2Iterator<float> yIt(message, "y");
-        sensor_msgs::PointCloud2Iterator<float> zIt(message, "z");
+        const auto positionField = results.GetConstFieldSpan<RaycastResultFlags::Point>().value();
+        auto positionIt = positionField.begin();
 
-        AZStd::optional<sensor_msgs::PointCloud2Iterator<float>> intensityIt;
+        AZStd::optional<sensor_msgs::PointCloud2Iterator<float>> messageIntensityIt;
+        AZStd::optional<RaycastResults::FieldSpan<RaycastResultFlags::Intensity>::const_iterator> intensityIt;
         if (isIntensityEnabled)
         {
-            intensityIt = sensor_msgs::PointCloud2Iterator<float>(message, "intensity");
+            messageIntensityIt = sensor_msgs::PointCloud2Iterator<float>(message, "intensity");
+            intensityIt = results.GetConstFieldSpan<RaycastResultFlags::Intensity>().value().begin();
         }
 
         const auto entityTransform = GetEntity()->FindComponent<AzFramework::TransformComponent>();
         const auto inverseLidarTM = entityTransform->GetWorldTM().GetInverse();
-        for (size_t i = 0; i < results.m_points.size(); ++i)
+        for (size_t i = 0; i < results.GetCount(); ++i)
```

Comment:
shouldn't we iterate using the `positionIt` iterator instead?

Reply:
When iterating over the results there may be many such iterators (one for position,  distance and/or intensity). Which of those should be used? `RaycastResults` guarantees that there is `GetCount()` entries of each result field, therefore we can safely use it.